### DISCUSSION TITLE
Determine if [TearDown] method is on the stack

### DISCUSSION THREAD

**Discussion by arnonax-tr** [nunit/nunit#4785] [Q&A]
I'm developing NuGets with common infrastructure for automation projects in our organization. As part of it, I have a method that should only be called from `[TearDown]`. In order to ensure that, I created a method that examines the current [`StackTrace`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stacktrace?view=net-8.0) and look for a method that has a `[TearDown]` attribute. 

This used to work perfectly before Net 8, but apparently due to new optimization, on some cases the `[TearDown]` method is optimized and replaced with a `InvokeStub_*` dynamically generated method, that apparently doesn't have the `[TearDown]` attribute.

I suppose that traversing the stack using the `StackTrace` object is not the best practice anyway, as methods on the stack can be in-lined or otherwise optimized (like in my case). 

My question is, is there a reliable way, or can you create one, to determine the current phase of the test? I.e., is it currently in One-time setup, test SetUp, Test body, tear down or one-time tear down. I looked both at `TestContext` and at `TestExecutionContext` and didn't find that information.

**Comment by stevenaw** [comment]
This is an interesting idea, though normally we would expect code which _must_ run in a given lifecycle phase to explicitly register itself via a mechanism like the `TearDown` attribute or the [ICommandBuilder interface](https://docs.nunit.org/articles/nunit/extending-nunit/ICommandWrapper-Interface.html).

It's tough to give specific guidance without knowing specifics about your situation. Can you explain a bit about the "common infrastructure" and the particular thing you are trying to only happen in the TearDown phase? 

Another option if your stack traces are changing on .NET 8 is to see if [`NoInlining`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.methodimploptions?view=net-8.0) may help.

**Reply by arnonax-tr** [reply]
Thanks! I'll look into the `ICommandBuilder` interface.
Regarding NoInlining, I don't think it can help as the `[TearDown]` method is not mine and I don't have control over it.

**Comment by OsirisTerje** [comment]
You can have a look at the work being done here, adding events for each step of the test process:  #4744.  It may be a way of solving this, but you then need handlers to detect the phase.  If all you want is to be ensured they are called, this could be the way.  

Otherwise, the best way is as @stevenaw to explicitly register your common infrastructure code.  The easiest way to do that is to simply make your common code a test base class, and let your code run as a TearDown method there.  NUnit states that the base class TearDown methods will be run last, so it will have to match that, but it **will** be run, so you know that it has been called. 
Going further, @stevenaw suggestion of ICommandBuilder is the way to go.

**Reply by arnonax-tr** [reply]
Thanks, I'll look into that link and at `ICommandBuilder`.

**Comment by arnonax-tr** [comment]
Hi again @OsirisTerje and @stevenaw . 

I looked into `ICommandWarpper` interface (which I actually use for something else). The problem with it is that it requires my clients to use a special attribute on _all_ of their test methods. This is not acceptable in my scenario for two reasons:
* There are many tests in many projects. Going through all of them and adding the new attribute is a lot of work (and even a simple find & replace won't work in all cases)
* It's easy to forget to add this attribute on new tests

The solution suggested in #4744 looks promising, but AFAIU is not ready yet.

I'll try to explain here my use case, though it's pretty complicated. I hope I'll be able to explain it clearly enough.

First, as a general rule, for every solution to a problem I develop, I do my best to make it as modular as possible (following SOLID principles), and avoid unnecessary dependencies. For example, many of the NuGet packages I provide are not nUnit dependent and may work with other frameworks as well (even if currently it's not strictly required). Only where necessary I create a NuGet package that depends on nUnit. Similarly, I have NuGet packages that depend on Selenium, others that depend on Extent Reports, etc, but where possible, I provide NuGet packages that are agnostic to specific technology stack (except .Net...).

Here are the relevant classes and packages that are relevant to my need to determine if a method is called from a `[TearDown]` method:

## Cleanup Manager

So one of my main packages is called `CleanupManager`. It exposes the following API:
```c#
public class CleanupManager
{
    public void AddCleanupAction(Action action);
    public void Cleanup();
}
```
> I simplify things a bit here for clarity

The `AddCleanupAction` adds delegates to a stack collection during the test, and on cleanup (tear-down) it invokes them in reverse order. It also applies some error handling mechanism so that if one cleanup action fails, it still continues to the others, and reporting all of the errors at the end. The idea is that on tear-down you'd only call `Cleanup`, and all of the actual code that perform the tear-down is modularized and more closely coupled to where the test (or setup) acquires resources that needs to be released. As this concept is relevant for most testing frameworks, it's not dependent on nUnit.

## Logging

I have a bunch of NuGets related to Logging. One of them contains mainly interfaces, and no dependency on any specific technology, but I also have one that depend both on nUnit and on Extent Reports. Among other things, this package logs information about the status of the test on tear-down and marks the status of the test in Extent Report. The recommended approach to do that is by registering it as a cleanup action with `CleanupManager`.

## NUnitCleanupManager

This component was added later due to a problem in following scenario:
* A test added a cleanup action, but that action threw an exception
* The exception was aggregated inside `CleanupManager`, but wasn't thrown yet because there were more cleanup actions.
* The cleanup action of the `NUnitExtentReportsLogger` was called, but it didn't know that the cleanup failed, and marked the test as passed. This caused an inconsistency between the status that appears in Extent Report and the one eventually reported by nUnit.

In order to fix this issue, I extracted the error handling of `CleanupManager` to a `virtual` method, and created a new package `NUnitCleanupManager` that override it, and calls `TestExecutionContext.CurrentContext.CurrentResult.RecordTearDownException(exception)` immediately in case of an exception in a cleanup action, before continuing to the next cleanup action. This allows `NUnitExtentReportsLogger` to examine `TestContext.Result.Outcome.Status` and report the status correctly.

The thing is that `TestExecutionContext.CurrentContext.CurrentResult.RecordTearDownException(exception)` can only be called from tear-down, and therefore I added a validation in the beginning of `NUnitCleanupManager.Cleanup` to ensure that there's a method with a `[TearDown]` attribute on the stack, and throw a more descriptive error if not. This is where I use `StackTrace` class to determine if a method with a `TearDownAttribute` exists on the stack, which as mentioned, stopped working when migrated to .Net 8.

> Note: later I also created an nUnit based `TestBase` class that incorporated these classes, but I'd still like to keep it modular and to allow clients to take only what they need.

**Comment by OsirisTerje** [comment]
Can you create a very small repro of this?   Just a single test with a teardown method, calling some library method, and performing the stack walk you do.  That will make it easier for us to look into it. 

You can add the repro into the https://github.com/nunit/nunit.issues repo, or attach it here.

UPDATE:
I added a small repro here:  https://github.com/nunit/nunit.issues/tree/main/Issue4785/take1.   (You can add yours as a "Take2" folder)

I just walked the stack, hopefully a bit like you do, and using .net 8.   I used the string name of the attribute, to avoid having to reference  NUnit  from the utility library.

And, it works.  

So you must do something else that blocks this.  And then, we need a repro from you. :-)

**Reply by arnonax-tr** [reply]
The issue is only reproduced when the library is built in Release configuration, and even then it doesn't happen consistently. AFAIU .net 8 performs some optimizations _during runtime_, and that's why it's inconsistent. Anyway, from what I understand (and the [answer I received](https://stackoverflow.com/a/78884513/980991) to my [SO question](https://stackoverflow.com/questions/78883965/looking-for-teardownattribute-on-the-stack-is-broken-in-net-8)), looking examining the StackTrace is not a good idea anyway and it's not guaranteed to be reliable.

The question is what alternative I have with nUnit to determine that the currently executing thread is in the tear-down phase.

**Reply by OsirisTerje** [reply]
I ran it both in debug and release, worked in both cases, multiple runs.  But I understand you have intermittent crashes.

 You can build a test package of NUnit based off #4744 and check if that works for you.  

The TestContext, as you state doesn't tell you which phase you're in, as you found.

**Reply by arnonax-tr** [reply]
Is there an ETA for #4744 to be released?

**Reply by OsirisTerje** [reply]
It will probably be 4.3.  We're releasing 4.2 any day now, and it will not reach that one, but 4.3 should be good.  We're trying to shorten the release cadence, but probably talking >= 3 months.   We can however, put out a beta version earlier. Once it is merged to main, there will be an alpha version on myget.

**Comment by manfred-brands** [comment]
> The exception was aggregated inside CleanupManager, but wasn't thrown yet because there were more cleanup actions.

I understand the aggregation, but why not throw an `AggregateException` at the end of all clean-ups?
That way your class would work with all platforms as they all should catch and report your exception.
In case of NUnit when called from a `TearDown` it would then indirectly call `RecordTearDownException`
and if called outside `TearDown` it would call `RecordExeption` and fail the test.

**Reply by arnonax-tr** [reply]
Throwing an `AggreateException` is what I did at the beginning (and still do from plain `CleanupManager` as opposed to `NUnitCleanupManager`). As I explained, the problem is that the `NUnitExtenLogger` doesn't know that there was an exception on `TearDown` and mark the test in the report as passed, even though it failed in tear-down.
